### ISSUE TITLE
using task.assertAgent in a custom build task fails when unit testing

### ISSUE THREAD

**Issue by BiribiriJaNai** [Microsoft/azure-pipelines-task-lib#299]
## **Description**
Using the latest version of vsts-task-lib (2.1.0), whenever I call the task.assertAgent function in a custom task, I get the following error while unit testing: tl.assertAgent is not a function.

![image](https://user-images.githubusercontent.com/29836717/33218240-2000c276-d13c-11e7-8789-ea87ca24a4f2.png)

![image](https://user-images.githubusercontent.com/29836717/33218273-4817a568-d13c-11e7-9ab3-203b6334ad72.png)

## **Repro steps**
I used the [AndroidSigning](https://github.com/Microsoft/vsts-tasks/blob/master/Tasks/AndroidSigning/androidsigning.ts) task in [vsts-tasks](https://github.com/Microsoft/vsts-tasks) repo as an example
 - run `git clone https://github.com/Microsoft/vsts-tasks.git`
 - edit the androidsigning.ts file. Locate the `run` function and add `tl.assertAgent('2.115.0');` as the first statement in the `try` block. Save the file
 - run `npm install`
 - run `node make.js build --task AndroidSigning`
 - run `node make.js test--task AndroidSigning`

## **Environment**
 - OS: Windows Server 2016 DataCenter 64bits (Azure VM, up to date)
 - Node JS: 6.12.0 (also tried with the latest LTS 8.9.1)
 - npm: 3.10.10 (also tried with 5.5.1)
 - vsts-task-lib; 2.1.0
 - VSCode: 1.18.1

## **Trace file**
```
##vso[task.debug]agent.TempDirectory=undefined
##vso[task.debug]agent.workFolder=undefined
##vso[task.debug]loading inputs and endpoints
##vso[task.debug]loading INPUT_FILES
##vso[task.debug]loading INPUT_JARSIGN
##vso[task.debug]loading INPUT_ZIPALIGN
##vso[task.debug]loading VSTS_TASKVARIABLE_KEYSTORE_FILE_PATH
##vso[task.debug]loaded 4
##vso[task.debug]Agent.ProxyUrl=undefined
##vso[task.debug]Agent.CAInfo=undefined
##vso[task.debug]Agent.ClientCert=undefined
##vso[task.debug]task result: Failed
##vso[task.issue type=error;]TypeError: tl.assertAgent is not a function
##vso[task.complete result=Failed;]TypeError: tl.assertAgent is not a function
```

## **Execution log**
```
 Microsoft Windows [Version 10.0.14393]
(c) 2016 Microsoft Corporation. All rights reserved.

C:\dev\github\microsoft\vsts-tasks>node make.js build --task AndroidSigning
tsc tool:
Version 1.8.7
C:\dev\github\microsoft\vsts-tasks\node_modules\.bin\tsc
npm tool:
3.10.10
C:\Program Files\nodejs\npm

------------------------------------------------------------
Building: AndroidSigning
------------------------------------------------------------


Building module C:\dev\github\microsoft\vsts-tasks\Tasks\Common\securefiles-common


> npm install

> tsc --outDir C:\dev\github\microsoft\vsts-tasks\_build\Tasks\Common\securefiles-common --rootDir C:\dev\github\microsoft\vsts-tasks\Tasks\Common\securefiles-common

> copying module resources
copying *.ps1
copying *.psd1
copying *.psm1
copying *.sh
copying *.txt
copying icon.png
copying icon.svg
copying module.json
copying node_modules
copying package.json
copying Strings
copying task.json
copying task.loc.json
copying Tests

> npm install C:\dev\github\microsoft\vsts-tasks\_build\Tasks\Common\securefiles-common

> npm install

> tsc --outDir C:\dev\github\microsoft\vsts-tasks\_build\Tasks\AndroidSigning --rootDir C:\dev\github\microsoft\vsts-tasks\Tasks\AndroidSigning

> copying task resources
copying *.ps1
copying *.psd1
copying *.psm1
copying *.sh
copying *.txt
copying icon.png
copying icon.svg
copying module.json
copying node_modules
copying package.json
copying Strings
copying task.json
copying task.loc.json
copying Tests

Build successful


C:\dev\github\microsoft\vsts-tasks>node make.js test --task AndroidSigning
tsc tool:
Version 1.8.7
C:\dev\github\microsoft\vsts-tasks\node_modules\.bin\tsc
mocha tool:
2.3.3
C:\dev\github\microsoft\vsts-tasks\node_modules\.bin\mocha

> tsc --rootDir C:\dev\github\microsoft\vsts-tasks\Tests --outDir C:\dev\github\microsoft\vsts-tasks\_build\Tests

> copying ps test lib resources
copying **\@(*.ps1|*.psm1)

> mocha C:/dev/github/microsoft/vsts-tasks/_build/Tasks/AndroidSigning/Tests/L0.js C:\dev\github\microsoft\vsts-tasks\_build\Tests\L0.js


  AndroidSigning Suite

task.issue details omitted
task.complete details omitted
TRACE FILE: C:\dev\github\microsoft\vsts-tasks\_build\Tasks\AndroidSigning\Tests\L0AndroidSkipSignAlign.js.log
    1) Do not sign or zipalign if nothing is selected

  General Suite
    √ Find invalid task.json
    √ Find nested task.json (55ms)
    √ Find .js with uppercase
    √ Find unsupported demands
    √ Find unsupported runsOn
    √ Find invalid server Task
    √ Find invalid message key in task.json
    √ Find missing string in .ts
    √ Find missing string in .ps1/.psm1


  9 passing (346ms)
  1 failing

  1) AndroidSigning Suite Do not sign or zipalign if nothing is selected:

      AssertionError: task should have succeeded
      + expected - actual

      -false
      +true

      at Context.it (C:\dev\github\microsoft\vsts-tasks\_build\Tasks\AndroidSigning\Tests\L0.js:19:9)
      at callFnAsync (C:\dev\github\microsoft\vsts-tasks\node_modules\mocha\lib\runnable.js:306:8)
      at Test.Runnable.run (C:\dev\github\microsoft\vsts-tasks\node_modules\mocha\lib\runnable.js:261:7)
      at Runner.runTest (C:\dev\github\microsoft\vsts-tasks\node_modules\mocha\lib\runner.js:421:10)
      at C:\dev\github\microsoft\vsts-tasks\node_modules\mocha\lib\runner.js:528:12
      at next (C:\dev\github\microsoft\vsts-tasks\node_modules\mocha\lib\runner.js:341:14)
      at C:\dev\github\microsoft\vsts-tasks\node_modules\mocha\lib\runner.js:351:7
      at next (C:\dev\github\microsoft\vsts-tasks\node_modules\mocha\lib\runner.js:283:14)
      at Immediate.<anonymous> (C:\dev\github\microsoft\vsts-tasks\node_modules\mocha\lib\runner.js:319:5)
```

**Comment by stephenmichaelf** [comment]
@BiribiriJaNai I haven't tested this but I think you need to modify your mock answers. e.g. - In L0AndroidSkipSignAlign.ts you create TaskLibAnswers and setup responses for the function calls. You will need to add a configuration option for assertAgent to return true. You will have to do this for each mock setup.

**Comment by BiribiriJaNai** [comment]
@stephenmichaelf I tried this but `assertAgent` is not considered a command like `which`, `exist`, `checkPath` or `exec`. Or maybe I didn't do it the right way...

So I've been digging a little bit and I wondered why for example, a call to `task.setResourcePath` didn't raise any error, although it's also a function exposed by `task`.
![image](https://user-images.githubusercontent.com/29836717/33327882-d5770998-d458-11e7-95cd-58ab57e76baa.png)

I think I found the answer in the `mock-task.ts` file where `setResourcePath` is also declared.
![image](https://user-images.githubusercontent.com/29836717/33329659-4a5aac52-d45d-11e7-882f-2a0070a9f02f.png)

I tried to add an identical empty declaration for `assertAgent` in this file and it solved my problem.
Is it planned to add support to these methods (FYI `task.getVariables` also raises an error) in the task mock?

**Comment by damccorm** [comment]
I think this solution is pretty much in line with what we recommend in our [step by step walkthrough](https://github.com/Microsoft/azure-pipelines-task-lib/blob/9cedd04cc64c6b899b8b7826c54ec67c4682aa28/node/docs/stepbystep.md#success-test), though you can do it a little more cleanly by only overriding the assertAgent function:
```
tr.registerMockExport('mt', {
    assertAgent: function() {
        return true;
    }
});
```

I'm going to close since it seems like this is working for you, feel free to reopen if appropriate though.

**Comment by BiribiriJaNai** [comment]
@damccorm I tried your more cleanly version without any success, I still get the same error.
I also tried changing the signature of the method from your sample, because `assertAgent` takes a parameter and its return type is void, as declared in the `task.d.ts` file: 
```
/**
 * Asserts the agent version is at least the specified minimum.
 *
 * @param    minimum    minimum version version - must be 2.104.1 or higher
 */
export declare function assertAgent(minimum: string): void;
```

What I did is just a workaround. It doesn't solve the fact that this method is not mocked on your end.
